### ISSUE TITLE
Generic cached return value cannot be converted to declared value even when they are the same type

### ISSUE THREAD

**Issue by jzazo** [microsoft/pyright#6953]
**Describe the bug**
I declare a generic return type, and I have a provider of that same return type, but cached. Pyright raises an error. If I remove the caching no error is reported. Mypy does not declare an error.

**Code**
```python
from abc import abstractmethod
from functools import cache
from typing import Any, Generic, Protocol, TypeVar

from typing_extensions import NamedTuple

T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)

class ServiceId(NamedTuple, Generic[T]):
    id: str

class Provider(Protocol[T_co]):
    @abstractmethod
    def __call__(self) -> T_co:
        ...

class Servicer:
    _d: dict[str, Any]

    @cache  # commenting this line removes the error
    def get_service(self, service_id: ServiceId[T]) -> T:
        return self._d[service_id.id]

    def get_provider(self, service_id: ServiceId[T]) -> Provider[T]:
        return lambda: self.get_service(service_id)  # Expression of type "() -> T@get_service" cannot be assigned to return type "Provider[T@get_provider]"
```

The complete error
```python
Expression of type "() -> T@get_service" cannot be assigned to return type "Provider[T@get_provider]"
  Type "() -> T@get_service" cannot be assigned to type "() -> T@get_provider"
    Function return type "T@get_service" is incompatible with type "T@get_provider"
      Type "T@get_service" cannot be assigned to type "T@get_provider"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

**VS Code extension or command-line**
I am running as command-line. Vs-code pylance reports the same error.
Which version? 1.1.345

**Comment by erictraut** [comment]
I'm not convinced this is a bug in pyright. The problem is with the definition of `functools.cache` in the typeshed stubs. It unfortunately doesn't preserve the signature of the decorated function. Only the return type is preserved, but in this case the return type is a type variable that can no longer be solved when you call the decorated function.

**Comment by jzazo** [comment]
I have no idea, but are [this](https://github.com/python/typeshed/issues/6347), [this](https://github.com/microsoft/pyright/issues/2578) and [this](https://github.com/python/typing/discussions/946) related?

**Comment by erictraut** [comment]
After thinking about this more, I'm pretty convinced that pyright is doing the right thing here, so I'm going to close this as "won't fix". I think the definition of `functools.cache` is problematic in that it doesn't maintain the signature of the function it's decorating. If you'd like to see it changed, please report the issue to the maintainers of typeshed.
